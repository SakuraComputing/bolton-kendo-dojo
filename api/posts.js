const express = require('express');
const router = express.Router();
const mongoose = require('mongoose');
const passport = require('passport');


// Post model
const Post = require('./models/Post');

// Va;idation
const validatePostInput = require('../validation/post');

//@route GET api/posts/test
// @desc Test post route
// @access Private
router.get('/test', (req, res) => res.json({
    mmsg: 'Posts works'
}));

// @route GET api/posts
// @desc Get posts
// @access Public
router.get('/', (req, res) => {
    Post.find()
    .sort({ date: -1 })
    .then(posts => res.json(posts))
    .catch(err => res.status(404).json({ nopostsfound: "No posts found"}));
})

// @route GET api/posts/:id
// @desc Get post by id
// @access Public
router.get('/:id', (req, res) => {
    Post.findById(req.params.id)
    .then(post => res.json(post))
    .catch(err => res.status(404).json({ nopostfound: "No post found with that Id"}));
})


// @route POST api/posts
// @desc Create post route
// @access Private
router.post('/', passport.authenticate('jwt', { session: false }), (req, res) => {

    const { errors, isValid } = validatePostInput(req.body);

    // Check validation
    if(!isValid) {
        return res.status(400).json(errors);
    }

    console.log(req.body.text);

    const newPost = new Post({
        text: req.body.text,
        name: req.body.name,
        avatar: req.body.avatar, 
        user: req.user.id
    });

    newPost.save().then(post => res.json(post))
});

// @route DELETE api/posts/:id
// @desc Delete post by Id
// @access Private
router.delete('/:id', passport.authenticate('jwt', { session: false }), (req, res) => {
    Post.findById(req.params.id)
    .then(post => {
        if(post.user.toString() !== req.user.id) {
            return res.status(401).json({ notauthorised: "User not authorized "})
        }
        post.remove().then(() => res.json({ success: true }))
        .catch(err => res.statusMessage(404).json({ postnotfound: "No post found "}))
    })
})


module.exports = router;