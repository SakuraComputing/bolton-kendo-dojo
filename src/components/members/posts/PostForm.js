import React, { Component } from 'react';
import { connect } from 'react-redux';
import TextAreaFieldGroup from '../../../common/TextAreaFieldGroup';
import { addPost } from '../../../actions/postActions';

export class PostForm extends Component {

    constructor(props) {
        super(props)
        this.state = {
            text: '',
            errors: {}
        }
    }

    onChange = (e) => {
        this.setState({ text: e.target.value })
    }
    
    render() {

        const { errors } = this.state;

        return (
            <div>
                <form onSubmit={this.onSubmit}>
                    <div className="form-group">
                                <TextAreaFieldGroup
                                    className="post-text"
                                    placeholder="Create a post"
                                    name="text"
                                    value={this.state.text}
                                    onChange={this.onChange}
                                    error={errors.text}
                                />
                            </div>
                    <button type="submit" className="post-submit">
                        Submit
                    </button>                
                </form>
            </div>
        )
    }
}
const mapStateToProps = state => ({
    auth: state.auth,
    errors: state.errors
})
export default connect(mapStateToProps, { addPost })(PostForm);
