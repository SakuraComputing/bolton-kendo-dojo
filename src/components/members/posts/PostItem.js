import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { deletePost } from '../../../actions/postActions';

class PostItem extends Component {

    onDeleteClick = (id) => {
        this.props.deletePost(id);
    }

    render() {

        const { post, auth } = this.props;

        return (
            <div className="post-container">
                <div className="comment-outer">
                    <div className="comment-inner-1">
                        <a href="profile.html">
                            <img className="rounded-circle d-none d-md-block"
                                    src={post.avatar} alt=""/>
                        </a>
                        <br/>
                        <p className="avatar-name">{post.name}</p>
                        <br/>
                        <Link to={`/post/${post._id}`} className="item-comments button-small">
                            Comments
                        </Link>
                    </div>

                    <div className="comment-inner-2">
                        <p className="comment-text">{post.text}</p>
                    </div>
                    {/* {post.user === auth.user.id ? ( */}
                    <div className="comment-inner-3">
                        {post.user === auth.user.id ? (
                            <button 
                                onClick={this.onDeleteClick.bind(this, post._id)}
                                type="button"
                                className="button-delete button-small item-delete"
                            >
                            Delete Post
                            </button>

                        ) : null}   
                    </div>
                </div>
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    return { auth: state.auth }
};

export default connect(mapStateToProps, { deletePost } )(PostItem);
