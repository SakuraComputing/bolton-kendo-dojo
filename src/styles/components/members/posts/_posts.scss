.feed {
    margin: 0 2%;
}

.post-submit {
    margin-left: 10px;
}
.feed-container {
    margin: 0px 20px 10px 20px;
    border: solid 1px $blue-black;
    border-radius: 5px;
    padding: 1%;
    font-size: 110%;
}

.feed-container:hover {
    background-color: $light-blue;
}

.item-delete {
    float: right;
    margin-top: 0;
}