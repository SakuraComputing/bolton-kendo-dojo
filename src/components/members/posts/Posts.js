import React, { Component } from 'react';
import { connect } from 'react-redux';
import { getPosts } from '../../../actions/postActions';
import PostForm from './PostForm';
import Spinner from '../../../common/Spinner';
import PostFeed from './PostFeed';

export class Posts extends Component {

  componentDidMount() {
      this.props.getPosts();
  }
  
  render() {

    const { posts, loading } = this.props.post;

    let postContent;

    if(posts == null || posts === undefined || loading) {
        postContent = <Spinner />
    } else {
        postContent = <PostFeed posts={posts} />
    }

    return (
      <div>
        <PostForm />
        {postContent}
      </div>
    )
  }
}

const mapStateToProps = state => ({
    post: state.post
})
export default connect(mapStateToProps, { getPosts })(Posts);
