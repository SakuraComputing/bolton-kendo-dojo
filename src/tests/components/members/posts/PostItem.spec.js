import React from 'react';
import { shallow } from 'enzyme';
import { PostItem } from '../../../../components/members/posts/PostItem';

describe.skip('Post Item Component', () => {

    let postItem, props, mockDeletePost; 
    const post = jest.fn();
    
    
    beforeEach(() => {
        mockDeletePost = jest.fn();
        props = {
            name: 'test',
            deletePost: mockDeletePost
        }
        postItem = shallow(<PostItem post={post} {...props} />)
    })
    it('should render correctly', () => {
        expect(postItem).toMatchSnapshot();   
    });

    it('should display a delete button', () => {
        expect(postItem.find('.item-delete').exists()).toBe(true);
    });

    it('should display a comments button', () => {
        expect(postItem.find('.item-comments').exists()).toBe(true);
    });
    describe('when clicking the delete button', () => {
        beforeEach(() => {
            props = {
                name: 'test',
                
                deletePost: mockDeletePost
            };
        
            postItem = shallow(<PostItem post={post} {...props} />)
        })

        it('should call deletePost', () => {
            postItem.find('.item-delete').simulate('click');
            expect(mockDeletePost).toHaveBeenCalled();
        });
    });
});