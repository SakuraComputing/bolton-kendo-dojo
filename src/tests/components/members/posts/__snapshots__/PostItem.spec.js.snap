// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post Item Component should render correctly 1`] = `undefined`;
